connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java

```diff
@@ -188,22 +167,25 @@ public void addConnector(ConnectorConfig connConfig, ConnectorContext ctx) {
             throw new ConnectException("Connector with name " + connName + " already exists");
 
         final Connector connector = instantiateConnector(connClass);
+        WorkerConnector workerConnector = new WorkerConnector(connName, connector, ctx, lifecycleListener);
+
         log.info("Instantiated connector {} with version {} of type {}", connName, connector.version(), connClass.getName());
-        connector.initialize(ctx);
+        workerConnector.initialize();
         try {
-            connector.start(connConfig.originalsStrings());
+            workerConnector.start(connConfig.originalsStrings());
         } catch (ConnectException e) {
             throw new ConnectException("Connector threw an exception while starting", e);
         }
 
-        connectors.put(connName, connector);
+        connectors.put(connName, workerConnector);
 
         log.info("Finished creating connector {}", connName);
     }
 
     /* Now that the configuration doesn't contain the actual class name, we need to be able to tell the herder whether a connector is a Sink */
     public boolean isSinkConnector(String connName) {
-        return SinkConnector.class.isAssignableFrom(connectors.get(connName).getClass());
+        WorkerConnector workerConnector = connectors.get(connName);
+        return SinkConnector.class.isAssignableFrom(workerConnector.delegate.getClass());
```

Comment:
Wow, this is a pretty good catch. Seems pretty non-obvious to track down. I don't see an obvious way to make this a bigger issue if it went wrong (like it did here when we redirected to `delegate`), but probably worth asking if there's anything we could have done? Maybe assigning to a `WorkerConnector` variable is enough, but alternatively assigning `workerConnector.delegate` to a `Connector` variable would help, or even an assertion on the type?